cv32/bsp/handlers.S

```diff
@@ -23,13 +23,110 @@
 .global __no_irq_handler
 .global sw_irq_handler
 .global verification_irq_handler
+.global software_irq_handler
+.global timer_irq_handler
+.global external_irq_handler
+.global fast0_irq_handler
+.global fast1_irq_handler
+.global fast2_irq_handler
+.global fast3_irq_handler
+.global fast4_irq_handler
+.global fast5_irq_handler
+.global fast6_irq_handler
+.global fast7_irq_handler
+.global fast8_irq_handler
+.global fast9_irq_handler
+.global fast10_irq_handler
+.global fast11_irq_handler
+.global fast12_irq_handler
+.global fast13_irq_handler
+.global fast14_irq_handler
+.global fast15_irq_handler
+
+.weak software_irq_handler
+.weak timer_irq_handler
+.weak external_irq_handler
+.weak fast0_irq_handler
+.weak fast1_irq_handler
+.weak fast2_irq_handler
+.weak fast3_irq_handler
+.weak fast4_irq_handler
+.weak fast5_irq_handler
+.weak fast6_irq_handler
+.weak fast7_irq_handler
+.weak fast8_irq_handler
+.weak fast9_irq_handler
+.weak fast10_irq_handler
+.weak fast11_irq_handler
+.weak fast12_irq_handler
+.weak fast13_irq_handler
+.weak fast14_irq_handler
+.weak fast15_irq_handler
+
```

Comment:
Do you really want these to be weak references? Are you expecting that the user will link in alternative implementations that override these?

Weak references are easy to get wrong, they have unexpected semantics, notably with libraries in ELF. They also have different semantics in PE (the file format on Windows), where they are weak aliases. I try to avoid them, unless there is an explicit compelling use case that requires them.

In this case, for verification, I can't see such a use case. Surely verifying hardware should expect fixed software behavior?  If there is an intention to provide explicit handlers later, it should be by updates to this BSP to add them, which would not need a weak reference anyway.

If you are going down this route, you should also make `__no_irq_handler`, `sw_irq_handler` and `verification_irq_handler` weak as well.

Reply:
In interrupt testing I would like to assign a specific handler for each interrupt.  In the interrupt verification test I would prefer not to trust any internal CSRs for interrupt verification.  By installing a specific handler with a specific ID parameter I can be assured when I enter said handler that mcause and mip are correct.

If there is a better methodology for installing specific handlers in vectored mode I am open to it.  I am simply used to using weak handlers in embedded verification tests.  We commonly use this internally throughout our embedded verification codebase for our ARM-M products.

Agreed that all handlers should be weak for consistency if that's the path we choose eventually.